raml 1.0 Objecttype

tardyp opened the pull request:
This is copy of https://github.com/econchick/ramlfications/pull/3
which already has a fair amount of discussion in it. 
I am not so sure where we want to continue discussion, probably it is better there, as we'll get context.

Since yesterday I had little time to work on it, and just added a version of OrderedDict with more concise repr.

---
Remaining things todo in this PR:
- [x] String should inherit from Scalar Type

Remaining things todo in another PR:
- [ ] includes and libraries support
- [ ] final decision on wether we separate pure type declaration from parsing and data validation

tardyp commented:
added support for basic raml fragments (to help for the unit tests)
added support for data validation, as I think the validation part is too important to patch it later

don't know if you heard of reviewable, but it looks like an alternative to github for the review, especially for this PR, which will probably get big. There is a possibility to easily review changes between PR versions
https://reviewable.io/reviews/spotify/ramlfications/80

econchick commented on ramlfications/errors.py:
```diff
@@ -49,7 +49,7 @@ class InvalidSecuritySchemeError(BaseRAMLError):
 #####
 
-class LoadRAMLError(Exception):
+class LoadRAMLError(BaseRAMLError):
```
So the original purpose of `BaseRAMLError` was for errors related to _parsing_ a RAML file. It was not meant to be used for anything outside of `ramlfications/parser.py`, or now `ramlfications/parser/*`. 

I can see a case for having a base exception class for the RAMLfications package, that a base RAML parser error inherits as well as a loader error, invalid version error, etc.

tardyp replied:
OK, I created a BaseRAMLParserError for that

econchick commented on ramlfications/errors.py:
```diff
@@ -57,9 +57,32 @@ class LoadRAMLError(Exception):
+        self.value = value
+        if position_hint is None:
+            position_hint = []
+        BaseRAMLError.__init__(self, "{0}: {1}, but got: {2}".format(
```
Hmm I can't remember if this is old-style; what's beneficial about calling the parent class's `__init__` rather than something like `super(DataTypeValidationError, self).__init__(...)`?

tardyp replied:
yes, its a bad habbit from me, I dont find the super syntax intuitive at all, and prefer explicit parent calling

econchick commented on ramlfications/loader.py:
```diff
@@ -1,25 +1,21 @@
 
 
 RAMLHEADER = "#%RAML "
+SUPPORTED_FRAGMENT_TYPES = ("DataType",)
```
I'm assuming this is for stuff like `#%RAML 1.0 Library` and `#%RAML 1.0 DataType` etc. in file headers?

econchick commented on ramlfications/parser/main.py:
```diff
@@ -25,6 +26,8 @@
 
 from .parameters import create_param_objs
 
+from ramlfications.types import create_type
```
this should be before the `from .parameters` line

econchick commented on ramlfications/types.py:
```diff
@@ -0,0 +1,192 @@
+
+@type_class("string")
+@attr.s
+class StringType(BaseType):
```
Looking at the [spec diagram](http://docs.raml.org/specs/1.0/#raml-10-spec-overview), I feel like there should be four "base" types, `ObjectType`, `ArrayType`, `ScalarType`, and `ExternalType`, all of which inherit from `BaseType`. Then `ScalarType` would have `facets` and `enum`.  Then there would be `StringType` or `StringScalarType` that would inherit from `ScalarType`.

tardyp replied:
yes, this is something I need to think about a little bit more

econchick commented on ramlfications/types.py:
```diff
@@ -0,0 +1,192 @@
+    description     = attr.ib(default="", repr=False, convert=Content)
+    type            = attr.ib(default="string", repr=False)
+
+    def validate(self, validated_objet, position_hint=None):
```
Having these validate methods goes a bit against the already defined validation pattern. What is the purpose of having these attached to the object, rather than in `ramlfications/validate.py`?

tardyp replied:
it is really not about raml validation, but about data-validation. maybe I should call this method validate_data(). I think that maybe the associated test will help you to have more insight about what I mean.

econchick commented on ramlfications/utils/parser.py:
```diff
@@ -73,3 +75,16 @@ def __trait(item, **kwargs):
     return INH_FUNC_MAPPING[obj_type]
+
+
+def convert_caml_case(name):
```
nitpick: it's "camel" not "caml"

tardyp replied:
done

econchick commented on ramlfications/types.py:
```diff
@@ -0,0 +1,192 @@
+from ramlfications.errors import DataTypeValidationError
+from ramlfications.parameters import Content
+from ramlfications.utils.common import OrderedDict
+from ramlfications.utils.parser import convert_caml_case
```
I'd put imports from the same module on the same lines, and use relative imports e.g.:

```python
from six import iteritems, string_types, MAXSIZE

from .errors import DataTypeValidationError, UnknownDataTypeError
from .parameters import Content
from .utils.common import OrderedDict
from .utils.parser import convert_caml_case
```

tardyp replied:
yeah, we have a different policy with buildbot, I can use your policy if you whish

Doing one line per import allows to ease up a lot with the merge conflicts, when two people insert a new import from the same module.

We even have an automated tool for splitting the import in a precommit hook:
https://github.com/buildbot/buildbot/blob/master/common/fiximports.py

econchick commented on .gitignore:
```diff
@@ -30,3 +30,4 @@ coverage.xml
 docs/_build/
 docs/_templates/
 cover
+sandbox
```
curious, where does this come from?

econchick replied:
and also is it something you can put in your global git config? not sure if it's project-related.

tardyp replied:
indeed, this is specific to my workflow

econchick commented on ramlfications/types.py:
```diff
@@ -0,0 +1,192 @@
+    :param <any> default: default value for this property
+    :param <BaseType> data_type: type definition of the property
+    """
+    required        = attr.ib(default=None, repr=False)
```
Default should probably be `False` here.

econchick commented on ramlfications/raml.py:
```diff
@@ -76,6 +78,12 @@ class RootNode(object):
+    # @todo:
+    # eventually, we should probably make a separate class for 0.8 and 1.0
+    # as some attributes are not exclusively present in one or the other spec
+    types            = attr.ib(repr=False, validator=root_types)
```
I think it's okay to go ahead and create a BaseRootNode object, and two child objects, one for 0.8 and one for 1.0. I'm wondering how to handle future-proofing, though... would making some "mixins" work?

tardyp replied:
I did this request in a separate PR

econchick commented:
Hey @tardyp -

I'd still like to separate out the functions you are using to parse the type objects from the object definition itself. I would put the `BaseType`, `ObjectType` etc definitions in `ramlfications/raml.py` since that module is meant for defining objects in the "first level" of a RAML file, e.g. traits, resource types, resource nodes.  I'm open to doing something like `ramlfications/raml/` with `raml.py` and `types.py` (and then `annotations.py` etc).  The only blurry line is that the built-in scalar types, e.g. string/number/integer types  are very similar to `ramlfications/parameters.py` with some similar/same attributes.  I'd then put the `parse_properties`, `create_property` and `create_type` in a separate module within `ramlfications/parser/`, or even within `ramlfications/parser/parameters.py`

I really do not like how `api.types` would return a dictionary. It's not congruent with the rest of the package's API, where it just returns a list of the objects.

I'm hesitant about the logic in `ramlfications/parser/__init__.py` & `ramlfications/loader.py`. Is ever a time where a RAML file would _just_ be a `DataType` and not a part of a complete API definition? I had imagined that a "main" RAML file would `!include` a DataType RAML, rather than just be by itself.  And with that, there wouldn't need to be any special if/else logic when parsing, it would just be all in the complete raw dictionary.  Also, what happens with this logic when a data type is defined inline within one main RAML file?

With that, I would add tests for using the `!include` functionality as well as when types are defined inline.

There has also been this open-ended question of when something is _not_ defined, is it `None`, or a default value, or not a Python `None` but more of like a `Nothing` object.  For instance, if an Object data type is defined, and _no_ minimum properties are defined, is it `None`? or should it just be `0`? Or should we reflect that it was truly not defined with a `Nothing`-type of object?  I've always defaulted to the first one (unless the RAML spec specifically defines a default value, e.g. query/uri/form parameters will default to "string" types). With your definition, you have a default values defined, which is not congruent with the current API, and something I would be very hesitant to add (generally for lack of reliability; for instance, `MAXSIZE` would be different across different python implementations and systems). I've been thinking of implementing a `Nothing` object similar to how [attrs handles it](https://github.com/hynek/attrs/blob/master/src/attr/_make.py#L12). In which case, I think it's best to just return `None` if nothing is defined for right now, as that is how the API behaves elsewhere.

tardyp commented:
re-separating object definitions from parsing:
I understand your wish for this, for me the code is easier to read like this rather than having to go back and forth from one file to the other.

re-factorization with parameter:
this is something I haven't looked at yet, indeed this is something we should take care of, and also we should factorize with the annotation types you are working on as well.

re-dictionary:
api.types is now a list.

about using ramlfication to directly import a DataType file, I do plan to use it as a feature in order to implement data-validation.
I think also Library makes sense to be loaded independently from an actual API using it.
For me this makes a lot of sense to share data types library between projects.

I hadn't look at !include yet, and Libraries, and wanted to do so in another PR.

None vs Nothing makes a lot of sense. will do.

tardyp commented:
see PR description for remaining TODOs

tardyp commented:
pushed a new version addressing some of the comments. rebased over https://github.com/spotify/ramlfications/pull/83
so I had to squash the history in order to ease up rebasing

tardyp commented:
Added support for scalar, int and bool.
I think we can merge this like this, and continue the work in further PRs (including split bw model and logic if you really think it is more suitable)

econchick commented on ramlfications/utils/parser.py:
```diff
@@ -111,3 +111,16 @@ def __trait(item, **kwargs):
+    """convert CamlCase to under_score names to better match pep8 style
+
+    :param string name: camlCase name of a raml Property
+    :returns string: under_score version of the raml Property
```
Just one final nitpick: on L117 and L120 - s/caml/camel :)

tardyp replied:
duh! I can only plaid I was coherently wrong

tardyp commented:
@econchick  let me know if you need more work on that PR, else, fill free to shipit!

econchick commented:
@tardyp Ok I think you need to rebase from v0.2.0 branch now since I merged #83

tardyp commented:
@econchick as, this PR is already on top of #83, there is no need to rebase

econchick commented:
@tardyp unfortunately github isn't smart enough, and still shows that your first commit isn't merged, but will after a rebase.

tardyp commented:
here is the rebase!

econchick commented:
:heart: thank you @tardyp !
